# Ekhator Uwaila's Notebook

### Word Embeddings

- We'll be using the [spacy](https://www.shanelynn.ie/word-embeddings-in-python-with-spacy-and-gensim/) library for embeddings. 

In [17]:
#!pip install spacy

In [18]:
#pip install spacy

In [14]:
#pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.2.0/en_core_web_sm-2.2.0.tar.gz

In [4]:
import spacy

In [36]:
#from numpy.random import seed
#seed(1)
#from tensorflow import set_random_seed
#set_random_seed(2)

Run the following cell once, it downloads the relevant spacy embeddings. 

In [12]:
#!python -m spacy download en_core_web_sm

In [11]:
#!python -m spacy download en_core_web_sm --target=~/anaconda3/envs/NLP/lib/python3.9/site-packages --upgrade

In [5]:
nlp = spacy.load("en_core_web_sm")

In [3]:
# create sentence.
sentence = nlp('The grass is green .')

# now check out the embedded tokens.
for token in sentence:
    print(token.text, token.vector.shape, token.vector)

The (96,) [ 1.1176367  -0.6485975  -0.5305207   2.6813617  -0.60251266 -0.39047047
 -0.7779956   0.06361836 -0.51140857  1.1415539   1.4621595   1.210716
 -0.7346289  -1.0209426  -0.92507267 -0.62145746 -0.9593326   2.3750756
 -1.3438724   0.33913434 -0.591809   -0.12916107 -0.05317664 -1.6779104
  1.7667464   0.06753059  1.3158545   0.30678707  0.02242778  0.35546002
  0.28812993 -0.5579191   1.1736645  -2.317893    0.26374075  1.2440248
  0.33035815 -1.1292491   0.25046095  4.293695   -0.24713266  1.2683114
 -1.2641441   0.7981607  -1.037181    0.40011275 -0.18160012  2.0641277
  1.5894725   0.3470932  -0.9535838  -1.1725378   0.49678475  0.5897087
  0.69023705  0.17437099 -1.1555524   0.3945408  -0.95651937  0.35465586
 -0.92056453  0.22104093  0.309849   -0.07244819  0.7793205  -0.6125721
  0.07617447 -0.98683465 -0.9941983  -1.6516142  -0.07806996  0.87705696
 -0.3135129  -0.41650376 -0.7940915  -0.34402055 -0.43732265 -0.44075364
  0.72600687  0.3121772   0.02908312 -1.1000757  -

### Topic Modeling

- Given a document, determine the topic of the document
- For this task, we'll use the Brown corpus of texts accessible via NLTK

In [3]:
import nltk
nltk.download('brown')

[nltk_data] Downloading package brown to
[nltk_data]     /Users/uwailaekhator/nltk_data...
[nltk_data]   Package brown is already up-to-date!


True

In [3]:
from nltk.corpus import brown
# View the categories (genres) available in the Brown corpus
print("Categories (Genres) in the Brown Corpus:")
print(brown.categories())

Categories (Genres) in the Brown Corpus:
['adventure', 'belles_lettres', 'editorial', 'fiction', 'government', 'hobbies', 'humor', 'learned', 'lore', 'mystery', 'news', 'religion', 'reviews', 'romance', 'science_fiction']


In [6]:
# View the file ids available in a specific category (e.g., news)
category = 'adventure'
print("\nFile IDs in the '{}' category:".format(category))
print(brown.fileids(category))



File IDs in the 'adventure' category:
['cn01', 'cn02', 'cn03', 'cn04', 'cn05', 'cn06', 'cn07', 'cn08', 'cn09', 'cn10', 'cn11', 'cn12', 'cn13', 'cn14', 'cn15', 'cn16', 'cn17', 'cn18', 'cn19', 'cn20', 'cn21', 'cn22', 'cn23', 'cn24', 'cn25', 'cn26', 'cn27', 'cn28', 'cn29']


In [7]:
# View the words in a specific file
file_id = 'ca01'
print("\nWords in the file '{}' (category: '{}'):".format(file_id, category))
print(brown.words(file_id))


Words in the file 'ca01' (category: 'adventure'):
['The', 'Fulton', 'County', 'Grand', 'Jury', 'said', ...]


In [1]:
pip install ipywidgets widgetsnbextension pandas-profiling

Note: you may need to restart the kernel to use updated packages.


In [6]:

import numpy as np
from collections import defaultdict
import tqdm # tqdm displays a progress bar
from tqdm import tqdm_notebook as tqdm # tqdm is a nice process indicator 

category_vectors = []

cats = brown.categories()
    
# for each category
for cat in cats:
    print(cat)
    # grab all of the documents
    for fileid in tqdm(brown.fileids(categories=[cat])):
        sents = brown.sents(fileids=[fileid])
        sent_vecs = []
        for sent in sents:
            sent = ' '.join(sent)
            sent = nlp(sent)
            # grab all of the words, find their embedding, sum all embeddings
            word_sum = np.sum([tok.vector for tok in sent], axis=0) # why axis=0?
            # add the now summed embedding to the list for this category
            sent_vecs.append(word_sum)
        category_vectors.append((cat,np.sum(sent_vecs, axis=0)))
    

adventure


/var/folders/3t/lfj_tnbd7fd1t5jjhk9_0zmw0000gq/T/ipykernel_35607/1695457682.py:14: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for fileid in tqdm(brown.fileids(categories=[cat])):


  0%|          | 0/29 [00:00<?, ?it/s]

belles_lettres


  0%|          | 0/75 [00:00<?, ?it/s]

editorial


  0%|          | 0/27 [00:00<?, ?it/s]

fiction


  0%|          | 0/29 [00:00<?, ?it/s]

government


  0%|          | 0/30 [00:00<?, ?it/s]

hobbies


  0%|          | 0/36 [00:00<?, ?it/s]

humor


  0%|          | 0/9 [00:00<?, ?it/s]

learned


  0%|          | 0/80 [00:00<?, ?it/s]

lore


  0%|          | 0/48 [00:00<?, ?it/s]

mystery


  0%|          | 0/24 [00:00<?, ?it/s]

news


  0%|          | 0/44 [00:00<?, ?it/s]

religion


  0%|          | 0/17 [00:00<?, ?it/s]

reviews


  0%|          | 0/17 [00:00<?, ?it/s]

romance


  0%|          | 0/29 [00:00<?, ?it/s]

science_fiction


  0%|          | 0/6 [00:00<?, ?it/s]

In [9]:
import pandas as pd

keys,values=zip(*category_vectors) # unzip using a *

data = pd.DataFrame({'cat':keys,'vectors':values})

In [10]:
data[:30]

cat                                            vectors
0        adventure  [-391.54767, -252.42824, 64.12453, -457.9177, ...
1        adventure  [-208.12047, -293.2177, 26.995764, -318.43024,...
2        adventure  [-235.97108, -250.355, 75.03412, -482.97522, -...
3        adventure  [-175.13382, -250.06433, -39.461098, -243.2764...
4        adventure  [-299.08322, -166.47307, 72.89785, -428.13815,...
5        adventure  [248.29102, -266.11127, 101.59741, -217.71622,...
6        adventure  [-331.19736, -350.97147, 66.65898, -307.16693,...
7        adventure  [99.70201, -150.66492, 67.114105, -224.48262, ...
8        adventure  [-365.29047, -532.0815, 277.13153, -378.90637,...
9        adventure  [-446.13568, -480.46902, 95.30582, -414.78204,...
10       adventure  [113.65657, -331.23016, 134.4924, -133.46725, ...
11       adventure  [-459.92938, -416.2808, 45.672615, -419.51862,...
12       adventure  [-258.63443, -465.8481, 53.01596, -337.88852, ...
13       adventure  [-38.007893, -358.9045, 121.2813, -328.5989, -...
14       adventure  [-92.911865, -325.6785, 44.86077, -250.57304, ...
15       adventure  [-14.442227, -454.05734, 83.08552, -161.60358,...
16       adventure  [-377.35477, -447.19785, 88.83253, -389.75107,...
17       adventure  [-75.09315, -280.8026, -22.720549, -161.47804,...
18       adventure  [-166.26833, -324.82816, 48.002598, -262.1167,...
19       adventure  [-64.306984, -294.40704, 116.125656, -401.6674...
20       adventure  [-12.939863, -250.79115, -82.23416, -164.36568...
21       adventure  [129.99106, -303.56982, 110.65253, -256.09534,...
22       adventure  [41.132614, -191.99646, -23.596249, -255.23952...
23       adventure  [-529.4639, -497.87314, 153.13426, -300.96707,...
24       adventure  [0.28080988, -195.85162, 40.587452, -293.88684...
25       adventure  [-129.62474, -346.91818, 33.620464, -168.15091...
26       adventure  [-254.73676, -467.6661, 56.49347, -343.11166, ...
27       adventure  [2.8662195, -297.5427, -102.7322, -63.841732, ...
28       adventure  [-380.4301, -356.3609, 93.532364, -443.5138, -...
29  belles_lettres  [280.32632, -518.00775, -46.561462, 115.043884...

In [11]:
total = len(data)

#### compute the baselines

In [12]:
print('random baseline {}'.format(1.0/len(cat)))

print('most common baseline?')
for cat in cats:
    print(cat, len(data[data.cat==cat])/total)

random baseline 0.06666666666666667
most common baseline?
adventure 0.058
belles_lettres 0.15
editorial 0.054
fiction 0.058
government 0.06
hobbies 0.072
humor 0.018
learned 0.16
lore 0.096
mystery 0.048
news 0.088
religion 0.034
reviews 0.034
romance 0.058
science_fiction 0.012


#### split the data into train/test

In [13]:
test = data.sample(frac=0.1,random_state=200)
train = data.drop(test.index)

test.shape, train.shape 

((50, 2), (450, 2))

#### train a logistic regression classifier

In [14]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
le.fit(data.cat) 
X = [x for x in train.vectors]
y = le.transform(train.cat)

In [15]:
from sklearn.linear_model import LogisticRegression

In [16]:
clfr = LogisticRegression(multi_class='multinomial', solver='lbfgs')

In [17]:
clfr.fit(X,y)

/Users/uwailaekhator/anaconda3/envs/NLP/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(multi_class='multinomial')

#### evaluate 

In [18]:
from sklearn.metrics import accuracy_score

In [19]:
test_y = le.transform(test.cat)
test_X = [x for x in test.vectors]

score = accuracy_score(clfr.predict(test_X), test_y)
score

0.48

### Results

- GoogleNews-vectors-negative300.magnitude 0.4 (w2v)
- wiki-news-300d-1M.magnitude 0.56 (bert)
- glove.6B.300d.magnitude 0.52 (glove)

In [20]:
test.shape, train.shape 

((50, 2), (450, 2))

In [21]:
#Preparing train.cat for keras

from sklearn import preprocessing

#test = data.sample(frac=0.1,random_state=200)
#train = data.drop(test.index)

le = preprocessing.LabelEncoder() # convert to numerical categories
ohe = preprocessing.OneHotEncoder() # convert categories to distributions (i.e., 1-hot vectors)
le.fit(data.cat)
yTrain = le.transform(train.cat).reshape(-1, 1) # this is magic
ohe.fit(yTrain)
yTrain = ohe.transform(yTrain).todense()

XTrain = np.array([x for x in train.vectors])

XTrain.shape, yTrain.shape

((450, 96), (450, 15))

In [22]:
#Preparing test.cat for keras

yTest = le.transform(test.cat).reshape(-1, 1) # this is magic

yTest = ohe.transform(yTest).todense()

XTest = np.array([x for x in test.vectors])

XTest.shape, yTest.shape

((50, 96), (50, 15))

### Installing and Importing Keras and Tensorflow

In [187]:
# Seed value
seed_value= 5

# 1. Set the `PYTHONHASHSEED` environment variable at a fixed value
import os
os.environ['PYTHONHASHSEED']=str(seed_value)

# 2. Set the `python` built-in pseudo-random generator at a fixed value
import random
random.seed(seed_value)

# 3. Set the `numpy` pseudo-random generator at a fixed value
import numpy as np
np.random.seed(seed_value)

# 4. Set the `tensorflow` pseudo-random generator at a fixed value
import tensorflow as tf
tf.random.set_seed(seed_value)
# for later versions: 
# tf.compat.v1.set_random_seed(seed_value)

# 5. Configure a new global `tensorflow` session
from keras import backend as K
#session_conf = tf.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
#sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
#K.set_session(sess)
# for later versions:
session_conf = tf.compat.v1.ConfigProto(intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)
sess = tf.compat.v1.Session(graph=tf.compat.v1.get_default_graph(), config=session_conf)
tf.compat.v1.keras.backend.set_session(sess)


In [7]:
#!pip install TensorFlow==2.1.0

In [120]:
#import tensorflow as tf
#tf.random.set_seed('36')
#print(tf.__version__)

In [188]:
#!pip install keras
import keras as keras
#keras.utils.set_random_seed(1337)
#print(keras.__version__)

In [221]:
#import numpy as np
#seed = 7
#np.random.seed(seed)
from keras.models import Sequential
from keras.layers import Dense, Dropout

### Building our model

#### Using ReLU

In [190]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='ReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

In [191]:
# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

In [194]:
# Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5275 - accuracy: 0.1622
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5280 - accuracy: 0.1622
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5268 - accuracy: 0.1622
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5272 - accuracy: 0.1622
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5284 - accuracy: 0.1622
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5277 - accuracy: 0.1622
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5285 - accuracy: 0.1622
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5280 - accuracy: 0.1622
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5270 - accuracy: 0.1622
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5282 - accuracy: 0.1622
Epoch 11/50
45/45 [

In [195]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 2.5269 - accuracy: 0.1600
Test model loss: 2.526930809020996
Test model accuracy: 0.1599999964237213


ReLU provides a low accuracy because it does not perform well with a large number of negative inputs. I would be using LeakyReLU.

#### Using LeakyReLU

In [179]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

In [180]:
# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

In [185]:
# Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 60.4309 - accuracy: 0.4444
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 37.3277 - accuracy: 0.4956
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 31.8698 - accuracy: 0.5267
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 36.5781 - accuracy: 0.5067
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 30.7343 - accuracy: 0.5222
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 34.2151 - accuracy: 0.5356
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 33.2329 - accuracy: 0.5244
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 31.2614 - accuracy: 0.5467
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 31.0978 - accuracy: 0.5267
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 35.6681 - accuracy: 0.4889
Epoch 11/

In [186]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 110.1719 - accuracy: 0.4000
Test model loss: 110.17194366455078
Test model accuracy: 0.4000000059604645


LeakyReLU provided a higher accuracy

### Changing the number of nodes in the hidden layer:

#### Increasing the number of nodes to 70:

In [200]:
# Building the model
model = Sequential()
model.add(Dense(70, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 376.4279 - accuracy: 0.2089
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 142.4483 - accuracy: 0.3311
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 95.4098 - accuracy: 0.3489
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 66.8963 - accuracy: 0.3844
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 66.9036 - accuracy: 0.3911
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 72.6219 - accuracy: 0.3978
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 58.2947 - accuracy: 0.4333
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 69.8901 - accuracy: 0.3800
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 47.8046 - accuracy: 0.4267
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 42.5874 - accuracy: 0.4356
Epoch 1

In [201]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 110.5609 - accuracy: 0.3000
Test model loss: 110.56088256835938
Test model accuracy: 0.30000001192092896


#### Increasing the number of nodes to 100:

In [65]:
# Building the model
model = Sequential()
model.add(Dense(100, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 375.2850 - accuracy: 0.2267
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 140.9713 - accuracy: 0.3489
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 101.3098 - accuracy: 0.3889
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 85.1597 - accuracy: 0.3956
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 82.7936 - accuracy: 0.3622
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 71.4605 - accuracy: 0.4378
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 64.9865 - accuracy: 0.4511
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 71.3877 - accuracy: 0.4000
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 81.8505 - accuracy: 0.3822
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 67.2751 - accuracy: 0.4156
Epoch 

In [66]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 139.1455 - accuracy: 0.4000
Test model loss: 139.1454620361328
Test model accuracy: 0.4000000059604645


#### Increasing the number of nodes to 150:

In [111]:
# Building the model
model = Sequential()
model.add(Dense(150, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 546.1319 - accuracy: 0.1978
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 182.3414 - accuracy: 0.3222
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 109.2897 - accuracy: 0.3800
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 93.2619 - accuracy: 0.3756
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 103.9117 - accuracy: 0.3822
Epoch 6/50
45/45 [==============================] - 0s 2ms/step - loss: 102.5906 - accuracy: 0.4133
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 95.2217 - accuracy: 0.3844
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 83.1872 - accuracy: 0.3933
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 85.5117 - accuracy: 0.3933
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 67.0774 - accuracy: 0.4244
Epoc

In [112]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 193.4111 - accuracy: 0.3800
Test model loss: 193.4111328125
Test model accuracy: 0.3799999952316284


#### Increasing the number of nodes to 200:

In [131]:
# Building the model
model = Sequential()
model.add(Dense(200, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 2ms/step - loss: 603.5488 - accuracy: 0.2133
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 259.0693 - accuracy: 0.2933
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 144.2126 - accuracy: 0.3711
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 121.6080 - accuracy: 0.4289
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 134.6480 - accuracy: 0.3711
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 137.7523 - accuracy: 0.3600
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 83.4264 - accuracy: 0.4378
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 95.7447 - accuracy: 0.4311
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 97.5514 - accuracy: 0.4200
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 86.9853 - accuracy: 0.4378
Epo

In [132]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 321.9365 - accuracy: 0.3800
Test model loss: 321.93646240234375
Test model accuracy: 0.3799999952316284


#### Decreasing the number of nodes to 20:

In [83]:
# Building the model
model = Sequential()
model.add(Dense(20, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 168.6183 - accuracy: 0.1844
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 70.9232 - accuracy: 0.3222
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 44.1803 - accuracy: 0.3444
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 33.7823 - accuracy: 0.3578
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 33.2211 - accuracy: 0.3667
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 23.5643 - accuracy: 0.4156
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 25.6041 - accuracy: 0.4178
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 20.7616 - accuracy: 0.4222
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 18.3945 - accuracy: 0.4556
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 19.9617 - accuracy: 0.4133
Epoch 11

In [84]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 17.8074 - accuracy: 0.3400
Test model loss: 17.807418823242188
Test model accuracy: 0.3400000035762787


- Increasing the numbers of nodes, initially had no effect on the accuracy score, but when i kept increasing them, they reduced the models accuracy and increased the loss on the test data.
- Decreasing the number of nodes also reduced the accuracy score.

### Changing the activation function of the hidden layer:

#### Using ELU:

In [77]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='ELU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 180.7166 - accuracy: 0.2156
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 3.5900 - accuracy: 0.1578
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5657 - accuracy: 0.1711
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5577 - accuracy: 0.1444
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5528 - accuracy: 0.1378
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5808 - accuracy: 0.1689
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5658 - accuracy: 0.1356
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5617 - accuracy: 0.1422
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5686 - accuracy: 0.1844
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5698 - accuracy: 0.1511
Epoch 11/50
45/45

In [78]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 2.5460 - accuracy: 0.1600
Test model loss: 2.5459511280059814
Test model accuracy: 0.1599999964237213


#### Using TANH:

In [79]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='tanh', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 1s 1ms/step - loss: 2.7118 - accuracy: 0.1400
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5732 - accuracy: 0.1667
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5961 - accuracy: 0.1756
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5756 - accuracy: 0.1444
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5313 - accuracy: 0.1689
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5022 - accuracy: 0.1778
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 2.4715 - accuracy: 0.1867
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5048 - accuracy: 0.1667
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 2.4565 - accuracy: 0.1978
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 2.4412 - accuracy: 0.1444
Epoch 11/50
45/45 [

In [80]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 2.3444 - accuracy: 0.1000
Test model loss: 2.3444104194641113
Test model accuracy: 0.10000000149011612


#### Using SIGMOID:

In [81]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='sigmoid', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6884 - accuracy: 0.1133
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6177 - accuracy: 0.1422
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6060 - accuracy: 0.1400
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6271 - accuracy: 0.1467
Epoch 5/50
45/45 [==============================] - 0s 987us/step - loss: 2.6077 - accuracy: 0.1267
Epoch 6/50
45/45 [==============================] - 0s 972us/step - loss: 2.5988 - accuracy: 0.1711
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 2.5969 - accuracy: 0.1533
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6367 - accuracy: 0.1111
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 2.6284 - accuracy: 0.1622
Epoch 10/50
45/45 [==============================] - 0s 977us/step - loss: 2.6095 - accuracy: 0.1622
Epoch 11/50
4

In [82]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 2.6497 - accuracy: 0.1600
Test model loss: 2.6496505737304688
Test model accuracy: 0.1599999964237213


- LeakyReLU provided a higher accuracy than all other models because it performs better in networks with a large number of negative inputs, compared to other activation functions. And our embeddings have lots of negative numbers.

### Changing the Learning rate

#### Using 0.01:

In [85]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.01)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 1s 1ms/step - loss: 313.6273 - accuracy: 0.2133
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 119.0432 - accuracy: 0.3644
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 83.5229 - accuracy: 0.3444
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 55.3766 - accuracy: 0.4000
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 61.7992 - accuracy: 0.4133
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 69.9710 - accuracy: 0.4000
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 49.1513 - accuracy: 0.4178
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 39.6066 - accuracy: 0.4556
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 38.3134 - accuracy: 0.4244
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 28.6555 - accuracy: 0.4822
Epoch 1

In [86]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 52.8532 - accuracy: 0.4000
Test model loss: 52.853240966796875
Test model accuracy: 0.4000000059604645


#### Using 0.1:

In [87]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.1)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 5494.3306 - accuracy: 0.1733
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 2605.0364 - accuracy: 0.2689
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 1404.7079 - accuracy: 0.3156
Epoch 4/50
45/45 [==============================] - 0s 957us/step - loss: 1271.5435 - accuracy: 0.2889
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 1758.3453 - accuracy: 0.3222
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 2835.8284 - accuracy: 0.2711
Epoch 7/50
45/45 [==============================] - 0s 902us/step - loss: 3380.7263 - accuracy: 0.2911
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 3333.9656 - accuracy: 0.3267
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 3041.6731 - accuracy: 0.3756
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 5070.6982 - ac

In [88]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 8485.5967 - accuracy: 0.3400
Test model loss: 8485.5966796875
Test model accuracy: 0.3400000035762787


#### Using 0.15:

In [89]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.15)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 12153.6885 - accuracy: 0.1733
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 6102.7446 - accuracy: 0.2778
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 5623.6626 - accuracy: 0.2844
Epoch 4/50
45/45 [==============================] - 0s 935us/step - loss: 4608.8784 - accuracy: 0.2733
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 6217.0078 - accuracy: 0.2956
Epoch 6/50
45/45 [==============================] - 0s 976us/step - loss: 7193.3789 - accuracy: 0.2933
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 10168.5488 - accuracy: 0.2978
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 6215.1021 - accuracy: 0.3578
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 6485.0068 - accuracy: 0.3378
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 10147.9121 -

In [90]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 17467.4395 - accuracy: 0.4400
Test model loss: 17467.439453125
Test model accuracy: 0.4399999976158142


#### Using 0.17:

In [139]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.17)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 980us/step - loss: 13018.5625 - accuracy: 0.1689
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 9798.6260 - accuracy: 0.2289
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 8283.7979 - accuracy: 0.2822
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 8952.1387 - accuracy: 0.3000
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 9058.8320 - accuracy: 0.3067
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 8043.8320 - accuracy: 0.2956
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 8876.3594 - accuracy: 0.2911
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 7954.9443 - accuracy: 0.3222
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 7483.6318 - accuracy: 0.3533
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 5085.7871 - acc

In [140]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 20753.4727 - accuracy: 0.3600
Test model loss: 20753.47265625
Test model accuracy: 0.36000001430511475


#### Using 0.001:

In [141]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 1s 2ms/step - loss: 242.6665 - accuracy: 0.2022
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 96.8732 - accuracy: 0.3156
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 73.9746 - accuracy: 0.3356
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 65.3389 - accuracy: 0.3600
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 50.7758 - accuracy: 0.3622
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 50.6060 - accuracy: 0.4067
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 46.9802 - accuracy: 0.3933
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 39.4215 - accuracy: 0.4089
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 37.8381 - accuracy: 0.4244
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 42.5586 - accuracy: 0.4267
Epoch 11

In [142]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 27.2034 - accuracy: 0.4600
Test model loss: 27.203428268432617
Test model accuracy: 0.46000000834465027


#### Using 0.0001:

In [143]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.0001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 1s 1ms/step - loss: 427.4035 - accuracy: 0.0689 
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 295.4301 - accuracy: 0.1067
Epoch 3/50
45/45 [==============================] - 0s 2ms/step - loss: 234.6976 - accuracy: 0.1333
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 199.9793 - accuracy: 0.1400
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 176.5047 - accuracy: 0.1778
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 158.1143 - accuracy: 0.1756
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 142.9853 - accuracy: 0.1978
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 129.9043 - accuracy: 0.2022
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 120.6527 - accuracy: 0.1867
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 112.0358 - accuracy: 0.217

In [144]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 53.6456 - accuracy: 0.2800
Test model loss: 53.645633697509766
Test model accuracy: 0.2800000011920929


#### Using 0.00001:

In [145]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.00001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=50, batch_size=10)

Epoch 1/50
45/45 [==============================] - 0s 1ms/step - loss: 659.0590 - accuracy: 0.0689
Epoch 2/50
45/45 [==============================] - 0s 1ms/step - loss: 613.6396 - accuracy: 0.0644
Epoch 3/50
45/45 [==============================] - 0s 1ms/step - loss: 575.7479 - accuracy: 0.0733
Epoch 4/50
45/45 [==============================] - 0s 1ms/step - loss: 542.9491 - accuracy: 0.0800
Epoch 5/50
45/45 [==============================] - 0s 1ms/step - loss: 516.5143 - accuracy: 0.0889
Epoch 6/50
45/45 [==============================] - 0s 1ms/step - loss: 493.7683 - accuracy: 0.1000
Epoch 7/50
45/45 [==============================] - 0s 1ms/step - loss: 473.7477 - accuracy: 0.1000
Epoch 8/50
45/45 [==============================] - 0s 1ms/step - loss: 454.8261 - accuracy: 0.1067
Epoch 9/50
45/45 [==============================] - 0s 1ms/step - loss: 437.4015 - accuracy: 0.1022
Epoch 10/50
45/45 [==============================] - 0s 1ms/step - loss: 420.8462 - accuracy: 0.1111

In [146]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 135.2530 - accuracy: 0.2600
Test model loss: 135.2529754638672
Test model accuracy: 0.25999999046325684


- Learning rate of 0.001 produced the highest accuracy of 46%.

### Changing the number of epochs:

#### Using 40:

In [147]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=40, batch_size=10)

Epoch 1/40
45/45 [==============================] - 1s 1ms/step - loss: 254.3453 - accuracy: 0.1778
Epoch 2/40
45/45 [==============================] - 0s 1ms/step - loss: 100.8080 - accuracy: 0.2689
Epoch 3/40
45/45 [==============================] - 0s 1ms/step - loss: 71.0486 - accuracy: 0.2867
Epoch 4/40
45/45 [==============================] - 0s 1ms/step - loss: 60.1709 - accuracy: 0.3178
Epoch 5/40
45/45 [==============================] - 0s 1ms/step - loss: 50.4871 - accuracy: 0.3644
Epoch 6/40
45/45 [==============================] - 0s 1ms/step - loss: 44.7986 - accuracy: 0.3667
Epoch 7/40
45/45 [==============================] - 0s 1ms/step - loss: 41.7701 - accuracy: 0.3689
Epoch 8/40
45/45 [==============================] - 0s 1ms/step - loss: 37.6748 - accuracy: 0.4178
Epoch 9/40
45/45 [==============================] - 0s 1ms/step - loss: 36.8222 - accuracy: 0.4244
Epoch 10/40
45/45 [==============================] - 0s 1ms/step - loss: 35.7764 - accuracy: 0.4089
Epoch 1

In [148]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 33.6004 - accuracy: 0.3800
Test model loss: 33.60038757324219
Test model accuracy: 0.3799999952316284


#### Using 70:

In [149]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=70, batch_size=10)

Epoch 1/70
45/45 [==============================] - 0s 1ms/step - loss: 208.3239 - accuracy: 0.1044 
Epoch 2/70
45/45 [==============================] - 0s 944us/step - loss: 82.2999 - accuracy: 0.2244
Epoch 3/70
45/45 [==============================] - 0s 1ms/step - loss: 60.3410 - accuracy: 0.2822
Epoch 4/70
45/45 [==============================] - 0s 1ms/step - loss: 52.0307 - accuracy: 0.3200
Epoch 5/70
45/45 [==============================] - 0s 1ms/step - loss: 38.9277 - accuracy: 0.3733
Epoch 6/70
45/45 [==============================] - 0s 1ms/step - loss: 39.8395 - accuracy: 0.3533
Epoch 7/70
45/45 [==============================] - 0s 1ms/step - loss: 34.0268 - accuracy: 0.4089
Epoch 8/70
45/45 [==============================] - 0s 1ms/step - loss: 31.8437 - accuracy: 0.4267
Epoch 9/70
45/45 [==============================] - 0s 1ms/step - loss: 30.0061 - accuracy: 0.4311
Epoch 10/70
45/45 [==============================] - 0s 1ms/step - loss: 34.3452 - accuracy: 0.4111
Epoch

In [150]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 30.5583 - accuracy: 0.3600
Test model loss: 30.55834197998047
Test model accuracy: 0.36000001430511475


#### Using 100:

In [155]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=100, batch_size=10)

Epoch 1/100
45/45 [==============================] - 0s 1ms/step - loss: 244.7026 - accuracy: 0.1467
Epoch 2/100
45/45 [==============================] - 0s 1ms/step - loss: 110.1053 - accuracy: 0.2222
Epoch 3/100
45/45 [==============================] - 0s 1ms/step - loss: 76.8692 - accuracy: 0.2800
Epoch 4/100
45/45 [==============================] - 0s 1ms/step - loss: 63.8266 - accuracy: 0.3178
Epoch 5/100
45/45 [==============================] - 0s 1ms/step - loss: 49.4641 - accuracy: 0.3489
Epoch 6/100
45/45 [==============================] - 0s 1ms/step - loss: 44.2999 - accuracy: 0.3711
Epoch 7/100
45/45 [==============================] - 0s 1ms/step - loss: 43.2146 - accuracy: 0.3933
Epoch 8/100
45/45 [==============================] - 0s 1ms/step - loss: 35.5348 - accuracy: 0.4244
Epoch 9/100
45/45 [==============================] - 0s 1ms/step - loss: 34.2363 - accuracy: 0.4200
Epoch 10/100
45/45 [==============================] - 0s 1ms/step - loss: 34.9782 - accuracy: 0.38

In [156]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 34.2113 - accuracy: 0.3800
Test model loss: 34.211273193359375
Test model accuracy: 0.3799999952316284


#### Using 150:

In [153]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=150, batch_size=10)



Epoch 1/150
45/45 [==============================] - 0s 1ms/step - loss: 288.5738 - accuracy: 0.1689
Epoch 2/150
45/45 [==============================] - 0s 1ms/step - loss: 109.1605 - accuracy: 0.2156
Epoch 3/150
45/45 [==============================] - 0s 1ms/step - loss: 74.1864 - accuracy: 0.2822
Epoch 4/150
45/45 [==============================] - 0s 1ms/step - loss: 56.7892 - accuracy: 0.3200
Epoch 5/150
45/45 [==============================] - 0s 1ms/step - loss: 46.4132 - accuracy: 0.4022
Epoch 6/150
45/45 [==============================] - 0s 1ms/step - loss: 41.3097 - accuracy: 0.3667
Epoch 7/150
45/45 [==============================] - 0s 1ms/step - loss: 37.0286 - accuracy: 0.4022
Epoch 8/150
45/45 [==============================] - 0s 1ms/step - loss: 35.0888 - accuracy: 0.4244
Epoch 9/150
45/45 [==============================] - 0s 1ms/step - loss: 29.9328 - accuracy: 0.4511
Epoch 10/150
45/45 [==============================] - 0s 1ms/step - loss: 32.2521 - accuracy: 0.42

In [154]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 29.8052 - accuracy: 0.5000
Test model loss: 29.80516815185547
Test model accuracy: 0.5


#### Using 200:

In [203]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=200, batch_size=10)


Epoch 1/200
45/45 [==============================] - 0s 1ms/step - loss: 240.0355 - accuracy: 0.1333 
Epoch 2/200
45/45 [==============================] - 0s 1ms/step - loss: 104.4157 - accuracy: 0.2400
Epoch 3/200
45/45 [==============================] - 0s 1ms/step - loss: 67.7072 - accuracy: 0.2733
Epoch 4/200
45/45 [==============================] - 0s 1ms/step - loss: 54.3848 - accuracy: 0.3289
Epoch 5/200
45/45 [==============================] - 0s 1ms/step - loss: 45.2556 - accuracy: 0.3644
Epoch 6/200
45/45 [==============================] - 0s 1ms/step - loss: 42.9920 - accuracy: 0.3822
Epoch 7/200
45/45 [==============================] - 0s 1ms/step - loss: 33.2952 - accuracy: 0.4222
Epoch 8/200
45/45 [==============================] - 0s 1ms/step - loss: 31.2893 - accuracy: 0.4222
Epoch 9/200
45/45 [==============================] - 0s 1ms/step - loss: 28.6999 - accuracy: 0.4689
Epoch 10/200
45/45 [==============================] - 0s 1ms/step - loss: 30.1792 - accuracy: 0.4

In [204]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 42.7173 - accuracy: 0.3400
Test model loss: 42.717288970947266
Test model accuracy: 0.3400000035762787


- Using 150 epochs produced the highest accuracy of 50%

### Adding additional Dense layers

#### Using 2 dense layers:

In [207]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(150, activation='LeakyReLU'))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=150, batch_size=10)

Epoch 1/150
45/45 [==============================] - 1s 3ms/step - loss: 83.4854 - accuracy: 0.2000
Epoch 2/150
45/45 [==============================] - 0s 2ms/step - loss: 35.5814 - accuracy: 0.3311
Epoch 3/150
45/45 [==============================] - 0s 2ms/step - loss: 23.1944 - accuracy: 0.3822
Epoch 4/150
45/45 [==============================] - 0s 2ms/step - loss: 17.1324 - accuracy: 0.4511
Epoch 5/150
45/45 [==============================] - 0s 2ms/step - loss: 14.7899 - accuracy: 0.4778
Epoch 6/150
45/45 [==============================] - 0s 2ms/step - loss: 11.9106 - accuracy: 0.5000
Epoch 7/150
45/45 [==============================] - 0s 2ms/step - loss: 10.2890 - accuracy: 0.5222
Epoch 8/150
45/45 [==============================] - 0s 2ms/step - loss: 10.0018 - accuracy: 0.5444
Epoch 9/150
45/45 [==============================] - 0s 2ms/step - loss: 9.2926 - accuracy: 0.5133
Epoch 10/150
45/45 [==============================] - 0s 2ms/step - loss: 8.7423 - accuracy: 0.5156
E

In [208]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 22.4129 - accuracy: 0.4600
Test model loss: 22.412931442260742
Test model accuracy: 0.46000000834465027


#### Using 3 dense layers:

In [212]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(70, activation='LeakyReLU'))
model.add(Dense(150, activation='LeakyReLU'))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=150, batch_size=10)

Epoch 1/150
45/45 [==============================] - 1s 2ms/step - loss: 83.9787 - accuracy: 0.2222
Epoch 2/150
45/45 [==============================] - 0s 2ms/step - loss: 30.3786 - accuracy: 0.3067
Epoch 3/150
45/45 [==============================] - 0s 1ms/step - loss: 17.1593 - accuracy: 0.3311
Epoch 4/150
45/45 [==============================] - 0s 1ms/step - loss: 10.0931 - accuracy: 0.3867
Epoch 5/150
45/45 [==============================] - 0s 1ms/step - loss: 9.6058 - accuracy: 0.4356
Epoch 6/150
45/45 [==============================] - 0s 2ms/step - loss: 8.0153 - accuracy: 0.4733
Epoch 7/150
45/45 [==============================] - 0s 2ms/step - loss: 6.4349 - accuracy: 0.5000
Epoch 8/150
45/45 [==============================] - 0s 1ms/step - loss: 4.9001 - accuracy: 0.5467
Epoch 9/150
45/45 [==============================] - 0s 2ms/step - loss: 4.0610 - accuracy: 0.5867
Epoch 10/150
45/45 [==============================] - 0s 3ms/step - loss: 3.6927 - accuracy: 0.5778
Epoch

In [213]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 4ms/step - loss: 19.5868 - accuracy: 0.3400
Test model loss: 19.586780548095703
Test model accuracy: 0.3400000035762787


#### Using 4 dense layers:

In [218]:
# Building the model
model = Sequential()
model.add(Dense(200, activation='LeakyReLU', input_shape=(96,)))
model.add(Dense(150, activation='LeakyReLU'))
model.add(Dense(100, activation='LeakyReLU'))
model.add(Dense(50, activation='LeakyReLU'))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=150, batch_size=10)

Epoch 1/150
45/45 [==============================] - 1s 2ms/step - loss: 62.6993 - accuracy: 0.2067 
Epoch 2/150
45/45 [==============================] - 0s 2ms/step - loss: 13.4105 - accuracy: 0.3378
Epoch 3/150
45/45 [==============================] - 0s 2ms/step - loss: 9.4560 - accuracy: 0.3533
Epoch 4/150
45/45 [==============================] - 0s 1ms/step - loss: 6.4238 - accuracy: 0.3756
Epoch 5/150
45/45 [==============================] - 0s 1ms/step - loss: 5.7397 - accuracy: 0.4022
Epoch 6/150
45/45 [==============================] - 0s 1ms/step - loss: 4.6433 - accuracy: 0.4022
Epoch 7/150
45/45 [==============================] - 0s 1ms/step - loss: 3.4769 - accuracy: 0.4689
Epoch 8/150
45/45 [==============================] - 0s 1ms/step - loss: 2.9520 - accuracy: 0.5444
Epoch 9/150
45/45 [==============================] - 0s 1ms/step - loss: 2.2184 - accuracy: 0.5400
Epoch 10/150
45/45 [==============================] - 0s 1ms/step - loss: 2.7238 - accuracy: 0.5289
Epoch 

In [219]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 11.2912 - accuracy: 0.4000
Test model loss: 11.291173934936523
Test model accuracy: 0.4000000059604645


### Adding droput regularization

In [222]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(96,)))
model.add(Dropout(0.50))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain, yTrain, epochs=150, batch_size=10)

Epoch 1/150
45/45 [==============================] - 1s 2ms/step - loss: 411.6301 - accuracy: 0.1133
Epoch 2/150
45/45 [==============================] - 0s 1ms/step - loss: 184.1452 - accuracy: 0.1933
Epoch 3/150
45/45 [==============================] - 0s 1ms/step - loss: 139.1366 - accuracy: 0.1978
Epoch 4/150
45/45 [==============================] - 0s 1ms/step - loss: 112.3051 - accuracy: 0.2067
Epoch 5/150
45/45 [==============================] - 0s 972us/step - loss: 110.1026 - accuracy: 0.2511
Epoch 6/150
45/45 [==============================] - 0s 990us/step - loss: 95.6742 - accuracy: 0.2733
Epoch 7/150
45/45 [==============================] - 0s 953us/step - loss: 95.3048 - accuracy: 0.2467
Epoch 8/150
45/45 [==============================] - 0s 1ms/step - loss: 79.6700 - accuracy: 0.2911
Epoch 9/150
45/45 [==============================] - 0s 1ms/step - loss: 72.9885 - accuracy: 0.3200
Epoch 10/150
45/45 [==============================] - 0s 1ms/step - loss: 72.2940 - accur

In [223]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest, yTest)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 4.3236 - accuracy: 0.5000
Test model loss: 4.32356071472168
Test model accuracy: 0.5


### Try a different embedding (i.e., from spacy)

#### Using word2vec:

In [224]:
pip install gensim

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 7.5 MB/s eta 0:00:0000:0100:01
Note: you may need to restart the kernel to use updated packages.


In [8]:
import gensim

In [9]:
from nltk.corpus import brown
import numpy as np
from collections import defaultdict
import tqdm # tqdm displays a progress bar
from tqdm import tqdm_notebook as tqdm # tqdm is a nice process indicator 

category_vectors2 = []

cats2 = brown.categories()
    
# for each category
for cat in cats2:
    print(cat)
    # grab all of the documents
    for fileid in tqdm(brown.fileids(categories=[cat])):
        sents = brown.sents(fileids=[fileid])
        sent_vecs = []
        for sent in sents:
            #sent = ' '.join(sent)
            #sent = nlp(sent)
            # Create CBOW model
            model = gensim.models.Word2Vec([sent], min_count=1, vector_size=100, window=5)
 
            # grab all of the words, find their embedding, sum all embeddings
            word_sum = np.sum([model.wv[tok] for tok in sent], axis=0) # why axis=0?
            # add the now summed embedding to the list for this category
            sent_vecs.append(word_sum)
        category_vectors2.append((cat,np.sum(sent_vecs, axis=0)))
    

adventure


/var/folders/3t/lfj_tnbd7fd1t5jjhk9_0zmw0000gq/T/ipykernel_35607/2002206695.py:15: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for fileid in tqdm(brown.fileids(categories=[cat])):


  0%|          | 0/29 [00:00<?, ?it/s]

belles_lettres


  0%|          | 0/75 [00:00<?, ?it/s]

editorial


  0%|          | 0/27 [00:00<?, ?it/s]

fiction


  0%|          | 0/29 [00:00<?, ?it/s]

government


  0%|          | 0/30 [00:00<?, ?it/s]

hobbies


  0%|          | 0/36 [00:00<?, ?it/s]

humor


  0%|          | 0/9 [00:00<?, ?it/s]

learned


  0%|          | 0/80 [00:00<?, ?it/s]

lore


  0%|          | 0/48 [00:00<?, ?it/s]

mystery


  0%|          | 0/24 [00:00<?, ?it/s]

news


  0%|          | 0/44 [00:00<?, ?it/s]

religion


  0%|          | 0/17 [00:00<?, ?it/s]

reviews


  0%|          | 0/17 [00:00<?, ?it/s]

romance


  0%|          | 0/29 [00:00<?, ?it/s]

science_fiction


  0%|          | 0/6 [00:00<?, ?it/s]

In [229]:
import pandas as pd

keys2,values2=zip(*category_vectors2) # unzip using a *

data2 = pd.DataFrame({'cat':keys2,'vectors':values2})

In [230]:
test2 = data2.sample(frac=0.1,random_state=200)
train2 = data2.drop(test2.index)

test2.shape, train2.shape 

((50, 2), (450, 2))

In [231]:
#Preparing train2.cat for keras

from sklearn import preprocessing

#test = data.sample(frac=0.1,random_state=200)
#train = data.drop(test.index)

le2 = preprocessing.LabelEncoder() # convert to numerical categories
ohe2 = preprocessing.OneHotEncoder() # convert categories to distributions (i.e., 1-hot vectors)
le2.fit(data2.cat)
yTrain2 = le2.transform(train2.cat).reshape(-1, 1) # this is magic
ohe2.fit(yTrain2)
yTrain2 = ohe2.transform(yTrain2).todense()

XTrain2 = np.array([x for x in train2.vectors])

XTrain2.shape, yTrain2.shape

((450, 100), (450, 15))

In [232]:
#Preparing test2.cat for keras

yTest2 = le2.transform(test2.cat).reshape(-1, 1) # this is magic

yTest2 = ohe2.transform(yTest2).todense()

XTest2 = np.array([x for x in test2.vectors])

XTest.shape, yTest.shape

((50, 96), (50, 15))

In [234]:
# Building the model
model = Sequential()
model.add(Dense(50, activation='LeakyReLU', input_shape=(100,)))
model.add(Dense(15, activation='softmax'))

# Compiling the model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#Train the model
model.fit(XTrain2, yTrain2, epochs=150, batch_size=10)

Epoch 1/150
45/45 [==============================] - 1s 2ms/step - loss: 3.2267 - accuracy: 0.1356
Epoch 2/150
45/45 [==============================] - 0s 2ms/step - loss: 2.5995 - accuracy: 0.1533
Epoch 3/150
45/45 [==============================] - 0s 1ms/step - loss: 2.4851 - accuracy: 0.1889
Epoch 4/150
45/45 [==============================] - 0s 1ms/step - loss: 2.4436 - accuracy: 0.1533
Epoch 5/150
45/45 [==============================] - 0s 1ms/step - loss: 2.4294 - accuracy: 0.1800
Epoch 6/150
45/45 [==============================] - 0s 1ms/step - loss: 2.3971 - accuracy: 0.1800
Epoch 7/150
45/45 [==============================] - 0s 2ms/step - loss: 2.3360 - accuracy: 0.1800
Epoch 8/150
45/45 [==============================] - 0s 2ms/step - loss: 2.3479 - accuracy: 0.1800
Epoch 9/150
45/45 [==============================] - 0s 2ms/step - loss: 2.3189 - accuracy: 0.1756
Epoch 10/150
45/45 [==============================] - 0s 1ms/step - loss: 2.3104 - accuracy: 0.1644
Epoch 11/

In [235]:
# Evaluating the model on the test data
loss, accuracy = model.evaluate(XTest2, yTest2)
print('Test model loss:', loss)
print('Test model accuracy:', accuracy)

2/2 [==============================] - 0s 3ms/step - loss: 2.3833 - accuracy: 0.2400
Test model loss: 2.3833048343658447
Test model accuracy: 0.23999999463558197


Word2vec produced a low accuracy of 23%

### Questions

What would you say is the neural network "learning"?
- The neural network is learning as it processes the labelled data (train dataset) inputed into it.

How does the depth or width of the network affect the training and the results?
- The depth has to do with the layers, while the width has to do with the node. An increase in the depth or width of the network, means allowing the network to capture more patterns in the data. However, one ha to be careful when increasing these 2 parameters, as it can lead to overfitting. <br>
In my codes above, increasing the width, reduced the accuracy score and increasing the depth, also reduced the accuracy score.

As you made changes to the network, what do you notice about how hyperparameters (network depth, number of nodes, learning rate, etc.) and how they interact with each other? <br>
- Increasing the numbers of nodes, initially had no effect on the accuracy score, but when i kept increasing them, they reduced the models accuracy and increased the loss on the test data.
Decreasing the number of nodes also reduced the accuracy score.
- LeakyReLU provided a higher accuracy than all other models because it performs better in networks with a large number of negative inputs, compared to other activation functions. And our embeddings have lots of negative numbers.
- Learning rate of 0.001 produced the highest accuracy of 46%, when i reduced it more, the accuracy decreased.
- Using 150 epochs produced the highest accuracy of 50%

We said that neural networks are learning non-convex problems, but what about finding the best parameters? <br> Is that a convex problem?
- Finding the best parameters for a neural network is a non-convex problem. The presence of multiple local minima and other complex structures in the optimization landscape makes it challenging to find the global minimum and also the best parameters.

What is regularization? Why is it important?
- Regularizatin is the process of simplifying an overly complex model.
- Regularization technique is important beacause it helps to prevent overfitting in a model. Examples of regularisatiom techniques are lasso, ridge, early stopping and dropout.

Which activation functions did you choose (besides logitistic/sigmoid)? For one of the activation functions you tried, spend some time learning about it. <br> Whereas logistic/sigmoid maps from inputs to a probability between 0-1, what does the activation function you chose do?

- After trying several activation functions: ReLU, ELU, TANH, SIGMOID AND LEAKYRELU, I decided to stick with LeakyReLU. LeakyReLU is similar to ReLU, but returns a small negative value for negative inputs rather than zero. Since our embeddings have a number of negative values, i felt this was a better choice.
- It has a small slope for negative values instead of zeroing thrm out liike ReLU.